mlflow/utils/name_utils.py

```diff
@@ -338,3 +368,264 @@ def _generate_random_name(sep="-", integer_scale=3, max_length=20):
     "youthful",
     "zealous",
 ]
+
+
+# Unique dataset names are generated from dataset digests by mapping the digest to an integer
+# and applying the modulo operation with a prime divisor to obtain an index. A prime divisor
+# is used to reduce the likelihood of collisions. This index is then mapped to a predicate
+# (e.g. "unique") and an integer in the range [0, 100000) (e.g. 9054). Finally, the dataset name
+# is constructed as <predicate>-data-<integer> (e.g. unique-data-9054)
+_GENERATOR_DATASET_DIGEST_MODULO_DIVISOR = 24999983
+_GENERATOR_DATASET_INTEGER_UPPER_BOUND = 10**5
+_GENERATOR_DATASET_PREDICATES = [
```

Comment:
Ran the following code and confirmed only 2 collisions for 10,000 digests, which seems acceptable:

```
from mlflow.utils.name_utils import _generate_dataset_name

dsn = []
for i in range(10000):
    dsn.append(_generate_dataset_name(str(i)))

print(len(dsn), len(set(dsn)))
```

```
$ python test.py
10000 9998
```

With 100,000 datasets, it goes up to ~ 200 collisions. This still seems acceptable. Another option is to put the digest in the name, but then we duplicate the digest and make the name longer; that might be okay. Thoughts @prithvikannan ?